### PR TITLE
Implement Backfill bitfinex with APIv2 & refactoring err.match function

### PR DESCRIPTION
**Pull request by nao0811ta**
fixed [#257](https://github.com/carlos8f/zenbot/issues/257)

### CODE REVIEW

File: commands/backfill.js [hunk 1]
```diff
@@ -66,13 +66,22 @@ module.exports = function container (get, set, clear) {
             var opts = {product_id: product_id}
             if (mode === 'backward') {
               opts.to = marker.from
+              // bitfinex v2 api required timestamp
+              if (exchange_id === 'bitfinex' && !marker.from){
```
Comment (DeviaVir) [inline commands/backfill.js:70]: We will really need to find a different solution for this, logic for exchanges should be limited to exchanges, and never live outside of that.

File: commands/backfill.js [hunk 2]
```diff
@@ -66,13 +66,22 @@ module.exports = function container (get, set, clear) {
...
             }
             else {
               if (marker.to) opts.from = marker.to + 1
               else opts.from = exchange.getCursor(start_time)
             }
             last_batch_opts = opts
-            exchange.getTrades(opts, function (err, trades) {
+            var tradeFunc = 'getTrades'
+            // for bitfinex v2 api for support historical api 
+            if (exchange_id === 'bitfinex'){
```
Comment (DeviaVir) [inline commands/backfill.js:81]: Same comment as before: this logic should happen in the exchange.js

File: extensions/exchanges/bitfinex/exchange.js [hunk 3]
```diff
@@ -22,24 +22,42 @@ module.exports = function container (get, set, clear) {
...
+    return public_client2
+  }
+ 
   function joinProduct (product_id) {
     return product_id.split('-')[0] + '' + product_id.split('-')[1]
   }
 
+  // symbol for v2 
+  function joinProduct2 (product_id) {
+    return 't' + product_id.split('-')[0] + '' + product_id.split('-')[1]
```
Comment (DeviaVir) [inline extensions/exchanges/bitfinex/exchange.js:38]: why not just `return 't' + joinProduct(product_id)`?

File: extensions/exchanges/bitfinex/exchange.js [hunk 4]
```diff
@@ -66,6 +84,48 @@ module.exports = function container (get, set, clear) {
...
+      args.sort = 1
+      // v2 limit is 1000
+      args.limit = 1000
+      if (opts.from) {
+        // initial value is null
+        args.start = opts.from
+      }
+      if (opts.to) {
+        // started point
+        args.end = opts.to
```
Comment (DeviaVir) [inline extensions/exchanges/bitfinex/exchange.js:101]: in the commands you are setting this to `new Date().getTime()`, why not just do that here?

**Review by DeviaVir — CHANGES_REQUESTED** [review]

**Comment by DeviaVir** [comment]
I also think there must be a nicer way than to use `function2` for the v2 things, maybe they should just live in different class?

File: commands/backfill.js (hunk 1 above)
Comment (nao0811ta) [reply commands/backfill.js:70]: Thank you
if opts.to is null , bitfinex response 1 years ago data.
hm.. I'm thinking..

File: commands/backfill.js (hunk 2 above)
Comment (nao0811ta) [reply commands/backfill.js:81]: got it..

File: extensions/exchanges/bitfinex/exchange.js (hunk 3 above)
Comment (nao0811ta) [reply extensions/exchanges/bitfinex/exchange.js:38]: I thought that. I forget to fix.

**Comment by nao0811ta** [comment]
@DeviaVir 
I'm sorry. and thank you, it must create new v2 class. but I don't have many time to survey v2 new API. so I only implemented backfill function for me ASAP. So, I hope that helps someone. 
@nedievas

**Comment by nao0811ta** [comment]
@nedievas  is working for v2
and my implementation is bad 
so close PR.

**Pull request closed without merging**